File: Leetcode_with_py/337rob.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
class Solution(object):
    def __init__(self):
        self.memo = {}

    def rob(self, root):
        if not root:
            return 0
        if root in self.memo:
            return self.memo[root]

        self.memo[root] = max(self.rob(root.left) + self.rob(root.right), \
        root.val + (0 if not root.left else self.rob(root.left.left) + self.rob(root.left.right)) + \
                    (0 if not root.right else self.rob(root.right.left) + self.rob(root.right.right)))
                    
        return self.memo[root]

File: Leetcode_with_Cpp/337rob.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    unordered_map<TreeNode*, int> memo;
    int rob(TreeNode* root) {
        if (!root) return 0;
        if (memo[root])
            return memo[root];
        int do_it = root->val + (root->left == NULL ? 0 : rob(root->left->left) + rob(root->left->right)) + 
                                (root->right == NULL ? 0 : rob(root->right->left) + rob(root->right->right));
        int not_do = rob(root->left) + rob(root->right);
        int res = max(do_it, not_do);
        memo[root] = res;
        return res;
    }
};